docs/design/2024-05-23-predicate-columns.md

```diff
@@ -0,0 +1,303 @@
+| ANALYZE TABLE t;                                 | DEFAULT(ALL)  | None       | None                                                                     | It will analyze all analyzable columns from the table.                   |
+| ANALYZE TABLE t LIST COLUMNS col1, col2;         | LIST          | col1, col2 | None                                                                     | It will only analyze col1 and col2.                                      |
+| ANALYZE TABLE t PREDICATE FOR PREDICATE COLUMNS; | PREDICATE     | None       | All predicate columns were collected before in mysql.column_stats_usage. | It will only analyze columns that exist in the mysql.column_stats_usage. |
+
+As you can see, we pick PREDICATE as the column_choice for the ANALYZE TABLE t PREDICATE COLUMNS statement. At the same time, we now consider DEFAULT to be ALL, but to support predicate columns during auto-analyze, we need to change the definition of DEFAULT.
+
+| Predicate Column Feature Status | Predicate Columns in `mysql.column_stats_usage` | Meaning                                                                                                                                                          |
+|---------------------------------|-------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Enabled                         | Present                                         | Use those predicate and indexed columns to analyze the table.                                                                                                    |
+| Enabled                         | Absent                                          | Only analyze indexed columns of the table. **If there are no indexes present, we will bypass the analysis of this table and directly set the modify_count to 0** |
+| Disabled                        | -                                               | Analyze all columns of the table                                                                                                                                 |
+
+After we change the definition of DEFAULT, we can use the predicate columns to analyze the table during auto-analyze.
+
+#### Analysis Dataflow
+
+![Dataflow](./imgs/auto-analyze-predicate-columns.png)
+
+### Cleanup Outdated Predicate Columns
+
+Users may have made schema changes, requiring the removal of non-existent columns from the `mysql.column_stats_usage` table.
+
+Before initiating the analyze process, we can first retrieve all predicate columns, compare them with the current schema, and remove any columns that no longer exist from the `mysql.column_stats_usage` table.
+
+### Global Variable
+
+In this feature, we introduce a new global variable `tidb_enable_column_tracking` to control whether to use predicate columns in the analyze process.
```

Comment:
Do you mean we add a new one and deprecate this one? Maybe `tidb_enable_only_analyze_predicate_coulumns`?

Reply:
@songrijie What do you think?

Reply:
Yes. Also maybe something like `tidb_analyze_default_column_choice`, which may be set to `predicate` or `all`.

Reply:
I agree with @time-and-fate that this is a good opportunity to reduce confusion about this variable.

Reply:
It's fine, but make it simple.  Does `tidb_analyze_column_option` make sense to you?

Reply:
So in this case, `tidb_enable_column_tracking` is deprecated since we will always track them. Is it correct?

Reply:
> So in this case, `tidb_enable_column_tracking` is deprecated since we will always track them. Is it correct?

Yes

Reply:
> It's fine, but make it simple. Does `tidb_analyze_column_option` make sense to you?

For me, `tidb_analyze_default_column_choice` makes more sense. Because `column_option` can have four different values: default, all, predicate, column_list. Here we only want to change the meaning of `default`, so I think `tidb_analyze_default_column_choice` makes more sense.

Reply:
LGTM

Reply:
Updated.